### DISCUSSION TITLE
Exploring Efficient Techniques for Handling Imbalanced Datasets in ML-For-Beginners

### DISCUSSION THREAD

**Discussion by Shaikhasna** [microsoft/ML-For-Beginners#890] [Q&A]
Hello everyone,

While working through the classification notebooks, I noticed that most examples assume relatively balanced datasets. In real-world scenarios, imbalanced datasets are common, and standard metrics like accuracy can be misleading.

I’m curious about:

The most effective strategies to integrate resampling methods (SMOTE, ADASYN, undersampling) within beginner-focused pipelines.

How to guide learners in choosing appropriate evaluation metrics like F1-score, ROC-AUC, or Matthews correlation coefficient without overwhelming them.

Best ways to demonstrate the impact of imbalance visually and practically in small-scale notebooks.

I’d love to hear thoughts from the community on balancing educational clarity with realistic ML practices. Sharing examples, tips, or alternative approaches would be highly appreciated.

**Comment by rohingosling** [comment]
While I am not an educator, I have written a number of user guides, and one runs into similar challenges related to balancing educational, or instructional clarity, with the messy and often verbose complexity of technical practice when writing user guides.

If I were to approach documenting strategies to target the ideas you mentioned, from the point of view of an instruction manual, or similar, then it might look something like this.

---

I've divided this response into two sections.

- One to explore ideas around managing the tension between educational clarity vs technical practice. 
- And the the second section, which covers your main question, in a style based on the educational clarity vs technical practice ideas in section 1.

---

## Part 1: A General Approach to Balancing Educational Clarity with Technical Practice

One way to resolve the tension between educational clarity and technical practice is to introduce each technique only after the student (or general audience) has personally hit the problem it solves. So you avoid laying out the full list of techniques up front, and instead let each tool show up as a fix for something that just broke. This idea gives you a three stage appraoch.

| Stage | Purpose | Student takeaway |
|---|---|---|
| 1. Awareness | Show the problem exists | *"Accuracy lied to me."* |
| 2. Diagnosis | Teach metrics that show the truth | *"Now I can see what the model is doing."* |
| 3. Treatment | Introduce techniques as fixes | *"I know which lever to pull, and why."* |

If it were an instruction manual, for example, then I would also build the whole chapter around a single running example (a synthetic 95/5 dataset, or a deliberately imbalanced version of an existing notebook, etc), so the audience only has to learn the data once, which is then used as a running use case throughout the material.

Kind of like how Microsoft used to use this fictional company, Northwind Traders, and their fictional database, the Northwind database, as the running use case for all their SQL Server and SQL Server Management Studio tutorials, etc. 

---

## Part 2: Applying the Staging to Your Questions

### Integrating Resampling Into Beginner Pipelines

The Order matters here:

1. **Class weights** (`class_weight='balanced'`). One-line change, with no new library, and no new concepts. This is what a working data scientist might try first, most of the time. Show the confusion matrix improving.
2. **Random undersampling.** Easy to explain, and lets you raise the trade-off (information loss) cleanly.
3. **SMOTE.** Now the audience can see what synthetic oversampling actually gives them. Plot the synthetic points in 2D so it isn't just a black box.
4. **ADASYN.** This can be framed as *"SMOTE that focuses on the hard region"*. Which is a refinement of a tool the audience already understands, as opposed to a parallel option.

Then, perhaps the single most important correctness point in this whole thing:

> Resample **inside** the cross-validation fold, never before the train/test split.

Resampling before splitting leaks synthetic samples into the test set and gives you fake good scores. It is the most common imbalance bug in beginner code.

One nice way to show people how to do this without being distracted by a long discussion about data leakage, is `imblearn.pipeline.Pipeline` from [imbalanced-learn](https://imbalanced-learn.org/). It is a drop-in replacement for sklearn's `Pipeline` that only resamples training folds. The code looks identical to what beginners already know, ...minus my personal OCD way of formatting code:

```python
from imblearn.pipeline       import Pipeline
from imblearn.over_sampling  import SMOTE
from sklearn.linear_model    import LogisticRegression
from sklearn.model_selection import cross_val_score

pipe = Pipeline (
    [
        ( 'smote', SMOTE              ( random_state = 42   )),
        ( 'clf',   LogisticRegression ( max_iter     = 1000 )),
    ]
)

scores = cross_val_score ( pipe, X, y, scoring = 'f1', cv = 5 )
```

That one substitution gives you correctness for free. As in, the `Pipeline` class itself enforces teh right behavior, regardless of whether the audience, or us for that matter, knows why it matters. If one resamples first, before the split, then it risks introducing a data leakage bug. 

### Choosing Metrics Without Overwhelming the Reader/Audience

Introduce metrics one at a time, each as the answer to a question the audience is already asking:

| Question the audience is already asking | Metric |
|:---|:---|
| "Of the cases I flagged, how many were real?" | Precision |
| "Of the real cases, how many did I catch?"    | Recall    |
| "How do I get one number from both?"          | F1        |
| "How well does my model rank cases overall?"  | ROC-AUC   |
| "What if the imbalance is severe?"            | PR-AUC    |

You could push **MCC** to an appendix or advanced section. It's a useful metric, but hard to motivate from scratch. F1 plus a confusion matrix gets you most of the way there with much less explanation. This could just be me being lazy, but, it's not only less writing for you, it's less reading for your audience.

The thing that can tie it all together is a **cost story**: medical screening (false negatives are bad) vs. spam filtering (false positives are annoying). Without that, *"pick the right metric"* feels arbitrary; with it, it's more obvious to the audience.

### Demonstrating Imbalance Visually

Three visuals, one per stage, should be all the tools the audience needs. The "stages" here are the "teaching instruction stages" I wrote about in the first part of this response, i.e., Awareness, Diagnosis, and Treatment.

| Visual | Stage | What it shows |
|---|---|---|
| 2D scatter of the two classes (PCA if needed) | 1 <br>Awareness | The minority class looks lonely. |
| Accuracy bar chart next to a confusion matrix heatmap | 2 <br>Diagnosis | Accuracy looks great; the matrix tells the truth. |
| Decision boundary, before vs. after class weighting or SMOTE | 3 <br>Treatment | The boundary visibly shifts. |

For SMOTE specifically, one can plot the original minority points and the synthetic ones in different colours. That turns the algorithm from a black box into something the audience can see.

---
<br>
So, those are my ideas.

**Comment by aasimansari1** [comment]
Great discussion on imbalanced datasets! One technique I found really useful is combining **SMOTE with ensemble methods** (like XGBoost with class_weight='balanced'). 

For interview prep on this topic, I've been building a resource: [ML Interview Prep](https://github.com/aasimansari1/ml-interview-prep) — it has a full section on handling imbalanced data with code examples. Might complement this curriculum well! ⭐ if helpful
